spec.bs

```diff
@@ -6,57 +6,417 @@ Status: CG-DRAFT
+
+In order to carefully control the flow of data read from Shared Storage, methods for getting data from the database backend are exposed to the {{SharedStorageWorklet}} only.
+
+Because bits of entropy can leak via {{SharedStorage/selectURL()}} if a user activates the fenced frame that navigates to the resulting opaque URL, we impose on each origin a <dfn>bit budget</dfn>, i.e. a maximum allowance of entropy bits that are permitted to leak during a given [=epoch=], where we define an <dfn>epoch</dfn> as a time period whose duration is a fixed pretermined length. When a leak occurs, its value in entropy bits is calculated and stored as a <dfn>bit debit</dfn> with a timestamp, and only [=bit debit=]s whose timestamps are within the current [=epoch=] are used to determine an origin's remaining budget. When an origin has insufficient remaining budget, {{SharedStorage/selectURL()}} will return its default value.
+
+<xmp class='idl'>
+  [Exposed=(Window,SharedStorageWorklet)]
+  interface SharedStorage {
+    [Exposed=(Window)] 
+    Promise<any> run(DOMString name,
+                     optional SharedStorageRunOperationMethodOptions options = {});
+
+    [Exposed=(Window)] 
+    Promise<USVString> selectURL(DOMString name,
                                  FrozenArray<SharedStorageUrlWithMetadata> urls,
                                  optional SharedStorageRunOperationMethodOptions options = {});
 
-  Promise<any> run(DOMString name, optional SharedStorageRunOperationMethodOptions options = {});
+    Promise<any> set(DOMString key,
+                     DOMString value,
+                     optional SharedStorageSetMethodOptions options = {});
+
+    Promise<any> append(DOMString key,
+                        DOMString value);
+
+    Promise<any> delete(DOMString key);
+
+    Promise<any> clear();
+
+    [Exposed=(SharedStorageWorklet)] Promise<DOMString> get(DOMString key);
+
+    [Exposed=(SharedStorageWorklet)] Promise<unsigned long> length();
+
+    [Exposed=(SharedStorageWorklet)] Promise<double> remainingBudget();
 
-  readonly attribute SharedStorageWorklet worklet;
-};
+    [Exposed=(SharedStorageWorklet)] SharedStorageIterator keys();
 
-[Exposed=(Window)]
-interface SharedStorageWorklet {
-  Promise<undefined> addModule(USVString moduleURL);
-};
+    [Exposed=(SharedStorageWorklet)] SharedStorageIterator entries();
 
-dictionary SharedStorageSetMethodOptions {
-  boolean ignoreIfPresent;
-};
+    [Exposed=(Window)] readonly attribute SharedStorageWorklet worklet;
+  };
 
-dictionary SharedStorageRunOperationMethodOptions {
-  object data;
-};
+  dictionary SharedStorageSetMethodOptions {
+    boolean ignoreIfPresent;
+  };
 
-dictionary SharedStorageUrlWithMetadata {
-  required USVString url;
-  object reportingMetadata;
-};
+  dictionary SharedStorageRunOperationMethodOptions {
+    object data;
+  };
+
+  enum IteratorMode { "key", "keyValue" };
+
+  [Exposed=(SharedStorageWorklet)]
+  interface SharedStorageIterator {
```

Comment:
This is interesting.  Are we using an iterator based on the idea that there can be too many keys/entries to load in a single invocation?  I haven't read fully, but we will need to define how these iterators work if storage is modified underneath them.

Also, it would probably be preferred to use the webidl async iterator feature.  For example, see the `async iterable<any>` line in the ReadableStream spec:

https://streams.spec.whatwg.org/#readablestream

Reply:
Done.